src/compiler/checker.ts

```diff
@@ -22964,13 +23055,177 @@ namespace ts {
             if (flags & TypeFlags.Union) {
                 return reduceLeft((type as UnionType).types, (facts, t) => facts | getTypeFacts(t, ignoreObjects), TypeFacts.None);
             }
+            if (flags & TypeFlags.Intersection) {
+                // // When an intersection contains a primitive type we ignore object type constituents as they are
+                // // presumably type tags. For example, in string & { __kind__: "name" } we ignore the object type.
+                ignoreObjects ||= maybeTypeOfKind(type, TypeFlags.Primitive);
+                return getIntersectionTypeFacts(type as IntersectionType, ignoreObjects);
+            }
+            return TypeFacts.All;
+        }
+
+        function getIntersectionTypeFacts(type: IntersectionType, ignoreObjects: boolean): TypeFacts {
+            let alwaysTrue = TypeFacts.None;
+            let alwaysFalse = TypeFacts.None;
+            let facts = TypeFacts.None;
+            for (const t of type.types) {
+                facts |= getTypeFacts(t, ignoreObjects);
+                const { alwaysTrue: tTrue, alwaysFalse: tFalse } = getAlwaysTypeFacts(t, ignoreObjects);
+                alwaysTrue |= tTrue;
+                alwaysFalse |= tFalse;
+            }
+
+            if (alwaysTrue & alwaysFalse) {
+                // Contradiction
+                return TypeFacts.None;
+            }
+
+            return (facts | alwaysTrue) & (~alwaysFalse);

```

Comment:
I think we don't need to compute `alwaysTrue` facts, since they are supposed to be a subset of `facts`. So right now they're just being used for detecting contradictions, as above, which I want to get rid of.